File: js/10-19/problem15.js
'use strict';
const prototypes = require('../prototypes');

// Finds the possible Lattice paths by imputting the dimensions.
var inputPathDimensions = function(x, y) {

	var surfaceArea = x + y;
	var d = function() {
		return (x > y) ? y : x
	}();

	var mul1 = [];
	var mul2 = [];

	for (let i = 0; i < d; i++) {
		mul1.push(surfaceArea - i);
		mul2.push(i + 1);
	}

	return mul1.product() / mul2.product();
};

console.log(inputPathDimensions(5, 3))
console.log(inputPathDimensions(20,20));

File: python/10-19/problem15.py
from functools import reduce


# Returns the product of a list.
def list_product(arr):
    return reduce(lambda x, y: x * y, arr)


def lattice_paths(x, y):
    surface_area = x + y
    d = y if x > y else x

    mul1 = [surface_area - x for x in range(d)]
    mul2 = [x + 1 for x in range(d)]

    return list_product(mul1) // list_product(mul2)


print(lattice_paths(20, 20))
